app/controllers/concerns/foreman/controller/registration_command.rb

```diff
@@ -0,0 +1,66 @@
+module Foreman::Controller::RegistrationCommand
+  extend ActiveSupport::Concern
+
+  private
+
+  def command
+    args_query = "?#{registration_args.to_query}"
+    "curl #{insecure} -X GET '#{endpoint}#{args_query if args_query != '?'}' #{command_headers} | bash"
+  end
+
+  def registration_args
+    args = registration_params.except(*ignored_query_args)
+
+    args['setup_insights'] = setup_insights.to_s
+    args['setup_remote_execution'] = setup_remote_execution.to_s
+
+    args.delete_if { |_, v| v.blank? }
+        .permit!
+  end
+
+  def insecure
+    registration_params['insecure'] ? '--insecure' : ''
+  end
+
+  def endpoint
+    return global_registration_url if registration_params['smart_proxy_id'].blank?
+
+    proxy = SmartProxy.authorized(:view_smart_proxies).find(registration_params['smart_proxy_id'])
+    "#{proxy.url}/register"
+  end
+
+  def host_config_params
+    organization = Organization.authorized(:view_organizations).find(registration_params['organization_id']) if registration_params['organization_id'].present?
+    location = Location.authorized(:view_locations).find(registration_params['location_id']) if registration_params['location_id'].present?
+    host_group = Hostgroup.authorized(:view_hostgroups).find(registration_params['hostgroup_id']) if registration_params["hostgroup_id"].present?
+    operatingsystem = Operatingsystem.authorized(:view_operatingsystems).find(registration_params['operatingsystem_id']) if registration_params["operatingsystem_id"].present?
+
+    Host.new(organization: organization, location: location, hostgroup: host_group, operatingsystem: operatingsystem).params
+  end
+
+  def setup_insights
+    return if registration_params['setup_insights'].to_s.blank?
+
+    from_host = host_config_params['host_registration_insights']
+    from_request = ActiveRecord::Type::Boolean.new.deserialize(registration_params['setup_insights'])
+
+    from_request if (from_request != from_host)
```

Comment:
Just to understand it right: this method returns one of 3 values: true, false, nil. If two of them mean the same (nil is considered as false, not a third different meaning) then why are here extra steps?... Moreover, if `from_request == from_host == true` then `nil` is returned which means `setup_insights() == false`, doesn't it?

Reply:
Yeah, the logic is confusing, and trust me, you are not the first one.
This will be completely removed with simpler behavior, see:

> As a user I want to be able to pick one of the three options for flags like REX, insights - force to yes, force to know, use the default value. The command should reflect that and either enforce the parameter value or avoid it entirely. https://bugzilla.redhat.com/show_bug.cgi?id=1919958